[test_sdpa_prefill.py] | Sanity tests nightly debug run with watcher | Sanity tests nightly debug run (Ubuntu 22.04 with watcher) / ttnn-unit-tests / ttnn sdpa group [wh_n300_civ2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] ttnn SDPA prefill sliding window streaming parity test fails: program size exceeds kernel config buffer on TENSIX
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** Sanity tests nightly debug run with watcher
**Job:** Sanity tests nightly debug run (Ubuntu 22.04 with watcher) / ttnn-unit-tests / ttnn sdpa group [wh_n300_civ2]

### Regression timeline

| | Commit | Date | Run URL |
|---|---|---|---|
| Last passing | N/A | N/A | N/A |
| First failing | `09a9c79f0980` | 2026-06-04T01:50:09Z | https://github.com/tenstorrent/tt-metal/actions/runs/26925019801 |
| Most recent failing | `e76a209ca567` | 2026-06-15T01:46:15Z | https://github.com/tenstorrent/tt-metal/actions/runs/27519220760 |

The regression has persisted for at least **11 days**.

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/27519220760/job/81398614578
- https://github.com/tenstorrent/tt-metal/actions/runs/27485044850/job/81275242492

### Failing test path(s)

- `tests/ttnn/unit_tests/operations/sdpa/test_sdpa_prefill.py::test_sdpa_sliding_window_streaming_parity[b=1-nh=8-nkv=1-s=1024-d=128-q_chunk_size=256-k_chunk_size=256-sliding_window=130-is_causal=False-bfp8]`

### Error excerpt (terminal failure from logs)

**Run 1 (2026-06-15):**
```
TT_FATAL: Program size (71472) too large for kernel config buffer (70656) on TENSIX  (assert.hpp:104)
RuntimeError: TT_FATAL @ /project/tt_metal/impl/program/program.cpp:2477: state.offset <= max_size
info:
Program size (71472) too large for kernel config buffer (70656) on TENSIX
FAILED tests/ttnn/unit_tests/operations/sdpa/test_sdpa_prefill.py::test_sdpa_sliding_window_streaming_parity[b=1-nh=8-nkv=1-s=1024-d=128-q_chunk_size=256-k_chunk_size=256-sliding_window=130-is_causal=False-bfp8] - RuntimeError: TT_FATAL @ /project/tt_metal/impl/program/program.cpp:2477: state.offset <= max_size
```

**Run 2 (2026-06-14):**
```
TT_FATAL: Program size (71472) too large for kernel config buffer (70656) on TENSIX  (assert.hpp:104)
RuntimeError: TT_FATAL @ /project/tt_metal/impl/program/program.cpp:2477: state.offset <= max_size
info:
Program size (71472) too large for kernel config buffer (70656) on TENSIX
FAILED tests/ttnn/unit_tests/operations/sdpa/test_sdpa_prefill.py::test_sdpa_sliding_window_streaming_parity[b=1-nh=8-nkv=1-s=1024-d=128-q_chunk_size=256-k_chunk_size=256-sliding_window=130-is_causal=False-bfp8] - RuntimeError: TT_FATAL @ /project/tt_metal/impl/program/program.cpp:2477: state.offset <= max_size
```

### Reproduction steps

On a `wh_n300` device (Wormhole N300), run:

```bash
python -m pytest tests/ttnn/unit_tests/operations/sdpa/test_sdpa_prefill.py::test_sdpa_sliding_window_streaming_parity \
  -k 'b=1-nh=8-nkv=1-s=1024-d=128-q_chunk_size=256-k_chunk_size=256-sliding_window=130-is_causal=False-bfp8' \
  -v
```

### Notes

The SDPA (Scaled Dot Product Attention) kernel for the sliding window streaming parity test is generating a program that is **816 bytes over the TENSIX kernel config buffer limit** (program size 71472 bytes vs. buffer limit 70656 bytes). This is a hard limit exceeded during program finalization/compilation (`program.cpp:2477`), not a numerical or precision failure.

The call stack confirms the failure path: `SDPAOperation` → `EnqueueMeshWorkload` → `MeshWorkloadImpl::compile` → `finalize_offsets` → `ProgramImpl::finalize_program_offsets`.

This may be caused by a kernel bloat regression — e.g., additional defines, larger loop bodies, or extra compute introduced in the sliding window non-causal (`is_causal=False`) SDPA kernel path around or after commit `09a9c79f0980`.

**Determinism check:**
- Run 1 terminal error: `TT_FATAL: Program size (71472) too large for kernel config buffer (70656) on TENSIX` (line 1174, 2026-06-15T10:39:04.0262302Z)
- Run 2 terminal error: `TT_FATAL: Program size (71472) too large for kernel config buffer (70656) on TENSIX` (line 1174, 2026-06-14T12:04:13.1339728Z)
- Both runs report **identical** program size (71472) and buffer limit (70656), on the same test variant with the same parameters.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: Sanity tests nightly debug run with watcher`
`Auto-triage-job-name: Sanity tests nightly debug run (Ubuntu 22.04 with watcher) / ttnn-unit-tests / ttnn sdpa group [wh_n300_civ2]`
<!-- AUTO-TRIAGE-METADATA-END -->

--- a/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_streaming.hpp
+++ b/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_streaming.hpp
@@ -41,6 +41,29 @@ struct MaybeProfileScope<true, timer_id> : kernel_profiler::profileScope<timer_i
 #define MaybeDeviceZoneScopedN(ENABLED, name)
 #endif
 
+// --- Outlined out-of-order pack (code-size) ---
+// pack_tile<true>() (absolute-address pack) inlines the full
+// llk_pack -> program_packer_destination GPR->FLOP address-programming sequence at every
+// call site. That inlined sequence is the single largest contributor to the PACK-thread
+// (TRISC2) text and overflows the TENSIX kernel-config buffer under watcher on Wormhole.
+// Outlining to one noinline copy trades a jal/ret per pack for a large code-size reduction.
+//
+// The jal/ret is NOT free on the hot inner loop. Empirically (wan2_2 BH perf check), the
+// softmax-exp pack in sub_exp_block_bcast_cols is the only perf-critical site: outlining it
+// alone caused the full regression (~70% -> ~66% math util), and re-inlining only it fully
+// recovers perf. So we keep that one site inlined via pack_tile<true>() directly and outline
+// everything else (output/SV drain, salad correction/sum, mask L1-accumulate) through this
+// wrapper. That keeps almost all of the Wormhole code-size win (only the exp pack's ~few
+// static copies return) with no measurable BH perf cost. On MATH/UNPACK threads pack_tile is
+// a no-op, so the outlined wrapper collapses to an empty inline function (zero overhead).
+#ifdef TRISC_PACK
+__attribute__((noinline, noclone)) static void sdpa_pack_tile_ooo(uint32_t dst, uint32_t cb, uint32_t idx) {
+    llk_pack<DST_ACCUM_MODE, true, PackMode::Default>(dst, cb, idx);
+}
+#else
+ALWI void sdpa_pack_tile_ooo(uint32_t, uint32_t, uint32_t) {}
+#endif
+
 static __attribute__((noinline, noclone)) void sdpa_cb_push_back_out_of_line(uint32_t cb_id, uint32_t num_tiles) {
     cb_push_back(cb_id, num_tiles);
 }
@@ -254,11 +277,11 @@ ALWI void pack_contiguous_rows_nocfg(
     for (uint32_t row = 0; row < row_count; ++row) {
         uint32_t out_tile_index = (row_base + row) * row_stride + col_base;
         if (use_blocked_pack_width) {
-            pack_tile<true>(dst_index, out_cb, out_tile_index);
+            sdpa_pack_tile_ooo(dst_index, out_cb, out_tile_index);
             dst_index += pack_width;
         } else {
             for (uint32_t col = 0; col < pack_width; ++col) {
-                pack_tile<true>(dst_index++, out_cb, out_tile_index + col);
+                sdpa_pack_tile_ooo(dst_index++, out_cb, out_tile_index + col);
             }
         }
     }
@@ -519,7 +542,7 @@ void sub_exp_block_bcast_cols(
                 }
 #pragma GCC unroll 1
                 for (uint32_t j = 0; j < tiles_per_column; ++j) {
-                    pack_tile<true>(dst_index++, reduce_cb, max_row_base + i);
+                    pack_tile<true>(dst_index++, reduce_cb, max_row_base + i);  // HOT: softmax exp, keep inline
                     if (global_col_base == 0 && j == 0) {
                         PACK((llk_pack_reconfig_l1_acc(1)));
                     }
@@ -639,7 +662,7 @@ void salad_correct_fused(
         if (fuse_sum_here) {
             configure_single_tile_pack(sum_out_cb);
             for (uint32_t i = 0; i < tiles_per_row; i++) {
-                pack_tile<true>(dst_index++, sum_out_cb, write_row_base + i);
+                sdpa_pack_tile_ooo(dst_index++, sum_out_cb, write_row_base + i);
             }
         }
         tile_regs_release();
@@ -654,7 +677,7 @@ void salad_correct_fused(
         tile_regs_wait();
         configure_single_tile_pack(sum_out_cb);
         for (uint32_t i = 0; i < tiles_per_row; i++) {
-            pack_tile<true>(i, sum_out_cb, write_row_base + i);
+            sdpa_pack_tile_ooo(i, sum_out_cb, write_row_base + i);
         }
         tile_regs_release();
     }
@@ -790,7 +813,7 @@ static inline void l1_acc_single_tile(uint32_t mask_cb, uint32_t tile_idx, uint3
     copy_tile(mask_cb, tile_idx, 0);
     tile_regs_commit();
     tile_regs_wait();
-    pack_tile<true>(0, out_cb, out_pos);
+    sdpa_pack_tile_ooo(0, out_cb, out_pos);
     tile_regs_release();
 }
 

